src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -0,0 +1,613 @@
+                            notePredecessor(p.WhenFalse);
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(node.Kind);
+                    }
+                }
+
+                hasPredecessor.Free();
+                return;
+
+                void notePredecessor(BoundDecisionDagNode successor)
+                {
+                    if (successor != null && !hasPredecessor.Add(successor))
+                    {
+                        GetDagNodeLabel(successor);
+                    }
+                }
+            }
+
+            protected new void Free()
+            {
+                _dagNodeLabels.Free();
+                base.Free();
+            }
+
+            protected virtual LabelSymbol GetDagNodeLabel(BoundDecisionDagNode dag)
+            {
+                if (!_dagNodeLabels.TryGetValue(dag, out LabelSymbol label))
+                {
+                    _dagNodeLabels.Add(dag, label = dag is BoundLeafDecisionDagNode d ? d.Label : _factory.GenerateLabel("dagNode"));
+                }
+
+                return label;
+            }
+
+            /// <summary>
+            /// A utility class that is used to scan a when clause to determine if it might assign a variable,
+            /// directly or indirectly. Used to determine if we can skip the allocation of pattern-matching
+            /// temporary variables and use user-declared variables instead, because we can conclude that they
+            /// are not mutated while the pattern-matching automaton is running.
+            /// </summary>
+            protected class WhenClauseMightAssignWalker : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private bool _mightAssignSomething;
+
+                public bool MightAssignSomething(BoundExpression expr)
+                {
+                    if (expr == null || expr.ConstantValue != null)
+                    {
+                        return false;
+                    }
+
+                    this._mightAssignSomething = false;
+                    this.Visit(expr);
+                    return this._mightAssignSomething;
+                }
+
+                public override BoundNode Visit(BoundNode node)
+                {
+                    // Stop visiting once we determine something might get assigned
+                    return this._mightAssignSomething ? null : base.Visit(node);
+                }
+
+                public override BoundNode VisitCall(BoundCall node)
+                {
+                    bool mightMutate =
+                        // might be a call to a local function that assigns something
+                        node.Method.MethodKind == MethodKind.LocalFunction ||
+                        // or perhaps we are passing a variable by ref and mutating it that way
+                        !node.ArgumentRefKindsOpt.IsDefault;
```

Comment:
Should we also check something like `!node.ArgumentRefKindsOpt.Any(rk => rk.IsWriteableReference())`?

Reply:
(it looks like this code has just been copied over without modification, but figured it was worth checking anyway.)

Reply:
This is a more conservative test.  I'm sure there are opportunities to make it less conservative, but I was just moving the code from one file to another here.

---
In reply to: [388680147](https://github.com/dotnet/roslyn/pull/42109#discussion_r388680147) [](ancestors = 388680147)

Reply:
I'm just moving the code from one file to another.  It is possible there is a correctness issue here.  I'm not confident one way or the other at this time - it may depend on whether this would apply to mutable value types.  Filed [https://github.com/dotnet/roslyn/issues/42300](https://github.com/dotnet/roslyn/issues/42300) to follow-up.

---
In reply to: [389136531](https://github.com/dotnet/roslyn/pull/42109#discussion_r389136531) [](ancestors = 389136531)